packages/vscode-integration/src/glsp-vscode-adapter.ts

```diff
@@ -0,0 +1,417 @@
+                    }
+                }
+            });
+        });
+
+        this.disposables.push(
+            this.diagnostics,
+            this.selectionUpdateEmitter,
+            serverMessageListener
+        );
+    }
+
+    /**
+     * Register a client on the GLSP-VSCode adapter. All communication will subsequently
+     * run through the VSCode integration. Clients do not need to be unregistered
+     * as they are automatically disposed of when the panel they belong to is closed.
+     *
+     * @param client The client to register.
+     */
+    public registerClient(client: GlspVscodeClient<D>): void {
+        this.clientMap.set(client.clientId, client);
+        this.documentMap.set(client.document, client.clientId);
+
+        // Set up message listener for client
+        const clientMessageListener = client.onClientSend(message => {
+            if (this.options.logging) {
+                if (isActionMessage(message)) {
+                    console.log(`Client (${message.clientId}): ${message.action.kind}`, message.action);
+                } else {
+                    console.log('Client (no action message):', message);
+                }
+            }
+
+            // Run message through first user-provided interceptor (pre-receive)
+            this.options.onBeforeReceiveMessageFromClient(message, (newMessage, shouldBeProcessedByAdapter) => {
+                if (shouldBeProcessedByAdapter) {
+                    this.processMessage(newMessage, 'client', (processedMessage, messageChanged) => {
+                        // Run message through second user-provided interceptor (pre-send) - processed
+                        const filteredMessage = this.options.onBeforePropagateMessageToServer(newMessage, processedMessage, messageChanged);
+                        if (typeof filteredMessage !== 'undefined') {
+                            this.options.server.onServerReceiveEmitter.fire(filteredMessage);
+                        }
+                    });
+                } else {
+                    // Run message through second user-provided interceptor (pre-send) - unprocessed
+                    const filteredMessage = this.options.onBeforePropagateMessageToServer(newMessage, newMessage, false);
+                    if (typeof filteredMessage !== 'undefined') {
+                        this.options.server.onServerReceiveEmitter.fire(filteredMessage);
+                    }
+                }
+            });
+        });
+
+        const viewStateListener = client.webviewPanel.onDidChangeViewState(e => {
+            if (e.webviewPanel.active) {
+                this.selectionUpdateEmitter.fire(this.clientSelectionMap.get(client.clientId) || []);
+            }
+        });
+
+        // Cleanup when client panel is closed
+        const panelOnDisposeListener = client.webviewPanel.onDidDispose(() => {
+            this.diagnostics.set(client.document.uri, undefined); // this clears the diagnostics for the file
```

Comment:
It might be usefull to use some sort of disposable collection here. Like it is done e.g in the custom editor example of VsCode (https://github.com/microsoft/vscode-extension-samples/blob/main/custom-editor-sample/src/dispose.ts)

Reply:
I understand the idea behind using such a collection but we cannot easily extend the webviewPanel type because it is defined by the vscode package. Additionally I believe implementing a dispose collection for 3 `.dispose()` calls would introduce more complexity than necessary.

If you insist however, I am happy to implement a collection like you described! I just wanted to counter-check your suggestion.